MinecraftForge/MinecraftForge
# [1.12.x] Custom Fluids Behave Like Water

Okay, this is the one I've been promising (threatening?) to post -- giving custom fluid system ability to hook into all water-like behavior.

**Background**: Custom fluids do not naturally do water-like things such as drown entities, push entities along with flow, water plants, float boats, etc. This PR addresses all these behaviors. The reason it hadn't been done previously is because Material.WATER was hard-coded throughout the vanilla code, so needs a patch for every behavior.

**Implementation Approach:**

- Material class has several added properties to reflect each waterlike behavior, along with setter and getter. Eg. setCanDrownEntity(). All default to false and existing liquids WATER and LAVA are set appropriately.

- Almost everywhere there is a check for Material.WATER was called, it is replaced with appropriate material property getter. The exceptions are anything to do with freezing (only water will turn to ice) and anything related specifically to BlockLiquid as that explicitly only handles WATER and LAVA.

- The isInsideMaterial() method is replaced with a ForgeHooks version that passes a predicate instead (for test against the material property). So you can test if you're inside a drownable material, a material that can float a boat, etc. Convenience predicates are included in ForgeHooks.

**Features Implemented**:

- Flowing fluids can be enabled to push entities like water.
- Fluids can be enabled to drown entities like water. Air supply HUD overlay and DROWN damage work same as water.
- Fluids can be enabled to float boats.
- Fluids can be enabled to water plants. This affects farmland and things like reed generation.
- Fluids can be enabled to spawn water creatures. 
- Fluids can be enabled to be "swimmable" meaning that any path navigation and AI will treat it as water note type.
- Fluids can be enabled to be absorbed by sponges.
- Fluids will extinguish fire.
- FOV changes like under water.

**Features Not Implemented** (Possible Future Enhancements):

- You can't fish in Fluids.
- Fluids don't freeze or otherwise relate to ice.
- Fluids' splash particles are same as water, might be cool to allow custom splash particles.
- If a sponge absorbs your fluid, you will still get water back from the sponge. But if modder doesn't like that just don't enable absorbing.
- Gases aren't really covered specifically, probably needs its own overhaul.

**Testing**:
- Make sure water still works as expected
- Place fluid with universal bucket (Misc creative tab) and try out all the various things you can do.

**Summary**:

I know this is a big PR, but the logic is pretty simple and it provides a LOT of functionality with good modder control.
